File: MathCore.DSP/Filters/EllipticLowPass.cs
using static System.Math;

using static MathCore.Polynom.Array;

// ReSharper disable InconsistentNaming

namespace MathCore.DSP.Filters;

/// <summary>Эллиптический фильтр нижних частот</summary>
public class EllipticLowPass : EllipticFilter
{
    /// <summary>Вычисляет порядок эллиптического фильтра</summary>
    /// <param name="dt">Период дискретизации</param>
    /// <param name="fp">Частота пропускания</param>
    /// <param name="fs">Частота заграждения</param>
    /// <param name="Gp">Затухание в полосе пропускания (по умолчанию -1 дБ)</param>
    /// <param name="Gs">Затухание в полосе заграждения (по умолчанию -40 дБ)</param>
    /// <returns>Порядок фильтра</returns>
    public static int GetOrder(double dt, double fp, double fs, double Gp = 0.891250938, double Gs = 0.01)
    {
        var kEps = Sqrt((1 / (Gp * Gp) - 1) / (1 / (Gs * Gs) - 1));
        var pi_dt = PI * dt;
        var kW = Tan(fs * pi_dt) / Tan(fp * pi_dt);

        var N = (int)Ceiling(T(kEps) * K(kW) / (K(kEps) * T(kW)));
        return N;
    }

    /// <summary>Вычисляет нули и полюса фильтра</summary>
    /// <param name="dt">Период дискретизации</param>
    /// <param name="fp">Частота пропускания</param>
    /// <param name="fs">Частота заграждения</param>
    /// <param name="Gp">Затухание в полосе пропускания (по умолчанию -1 дБ)</param>
    /// <param name="Gs">Затухание в полосе заграждения (по умолчанию -40 дБ)</param>
    /// <returns>Кортеж перечислений комплексных нулей и полюсов</returns>
    public static (IEnumerable<Complex> Zeros, IEnumerable<Complex> Poles) GetPoles(double dt, double fp, double fs, double Gp = 0.891250938, double Gs = 0.01)
    {
        var N = GetOrder(fp, fs, Gp, Gs);

        var EpsP = 1 / (Gp * Gp) - 1;
        var EpsS = 1 / (Gs * Gs) - 1;
        var zeros = EnumNormedZeros(N, EpsP, EpsS);
        var poles = EnumNormedPoles(N, EpsP, EpsS);

        var low_pass_zeros = TransformToLowPass(zeros, fp);
        var low_pass_poles = TransformToLowPass(poles, fp);

        var z_zeros = ToZ(low_pass_zeros, dt);
        if (N.IsOdd())
            z_zeros = z_zeros.AppendFirst(-1);
        var z_poles = ToZ(low_pass_poles, dt);

        return (z_zeros, z_poles);
    }

    /// <summary>Инициализирует коэффициенты передаточной функции эллиптического фильтра</summary>
    /// <param name="opt">Спецификация фильтра</param>
    /// <returns>Кортеж с коэффициентами полинома числителя и знаменателя</returns>
    private static (double[] A, double[] B) Initialize(Specification opt)
    {
        var k_W = 1 / opt.kW;
        var k_eps = 1 / opt.kEps;

        var N = (int)Ceiling(T(k_eps) * K(k_W) / (K(k_eps) * T(k_W))); // Порядок фильтра

        var eps_p = opt.EpsP;
        var (zeros, poles) = EnumNormedZerosPoles(N, eps_p, opt.EpsS);

        var z_poles = ToZArray(poles, opt.dt, opt.Wp);
        var z_zeros_enum = ToZ(zeros, opt.dt, opt.Wp);
        var is_odd = N.IsOdd();
        if (is_odd) z_zeros_enum = z_zeros_enum.AppendFirst(-1);
        var z_zeros = z_zeros_enum.ToArray();

        var k_zeros = z_zeros.Multiply(z => 1 - z);
        var k_poles = z_poles.Multiply(z => 1 - z);
        var k0 = is_odd 
            ? 1 
            : 1 / Sqrt(1 + eps_p * eps_p);

        var g_norm = k0 * (k_poles / k_zeros).Abs;

        var B = GetCoefficientsInverted(z_zeros).ToArray(b => b * g_norm).ToRe();
        var A = GetCoefficientsInverted(z_poles).ToRe();

        return (A, B);
    }

    /// <summary>Инициализирует новый эллиптический фильтр нижних частот</summary>
    /// <param name="dt">Период дискретизации</param>
    /// <param name="fp">Частота пропускания</param>
    /// <param name="fs">Частота заграждения</param>
    /// <param name="Gp">Затухание в полосе пропускания (по умолчанию -1 дБ)</param>
    /// <param name="Gs">Затухание в полосе заграждения (по умолчанию -40 дБ)</param>
    public EllipticLowPass(double dt, double fp, double fs, double Gp = 0.891250938, double Gs = 0.01)
        : this(GetSpecification(dt, fp, fs, Gp, Gs)) { }

    /// <summary>Инициализирует новый эллиптический фильтр нижних частот по спецификации</summary>
    /// <param name="Spec">Спецификация фильтра</param>
    public EllipticLowPass(Specification Spec) : this(Initialize(Spec), Spec) { }

    /// <summary>Инициализирует новый эллиптический фильтр</summary>
    /// <param name="Polynoms">Кортеж, содержащий массив коэффициентов полинома числителя и знаменателя</param>
    /// <param name="Spec">Спецификация фильтра</param>
    private EllipticLowPass((double[] A, double[] B) Polynoms, Specification Spec) : base(Polynoms.B, Polynoms.A, Spec) { }
}

File: MathCore.DSP/Filters/Python/filters/elliptic_lowpass.py
# Низкочастотный эллиптический фильтр (порт C# EllipticLowPass)
import numpy as np
from math import ceil, sqrt, pi, tan, atan

from .elliptic_filter import EllipticFilter
from .analog_based_filter import AnalogBasedFilter, Specification
from .digital_filter import DigitalFilter


class EllipticLowPass(EllipticFilter):
    """
    Цифровой низкочастотный эллиптический фильтр (ФНЧ).

    Обеспечивает минимальный порядок при заданных ограничениях на пульсации
    в полосе пропускания и ослабление в полосе заграждения.
    """

    @staticmethod
    def get_order(dt: float, fp: float, fs: float,
                  Gp: float = 0.891250938, Gs: float = 0.01) -> int:
        """Минимальный порядок эллиптического ФНЧ по формуле через K и K'."""
        k_eps = sqrt((1.0 / (Gp * Gp) - 1.0) / (1.0 / (Gs * Gs) - 1.0))
        pi_dt = pi * dt
        kW = tan(fs * pi_dt) / tan(fp * pi_dt)
        K = EllipticFilter.K
        T = EllipticFilter.T
        return int(ceil(T(k_eps) * K(kW) / (K(k_eps) * T(kW))))

    @staticmethod
    def _initialize(spec: Specification):
        """
        Коэффициенты ФНЧ эллиптического фильтра.

        Алгоритм
        --------
        1. N = ceil(T(1/kEps)·K(1/kW) / (K(1/kEps)·T(1/kW))).
        2. Нули и полюса нормированного эллиптического прототипа.
        3. Масштабируем на Wp и преобразуем билинейным методом.
        4. Нечётный порядок: добавляем нуль z=-1.
        5. Нормировка по DC: k0/|prod(1-z_zero)/prod(1-z_pole)|.
        """
        K = EllipticFilter.K
        T = EllipticFilter.T

        # обратные модули (LP-прототип: kW и kEps перевёрнуты относительно BS)
        k_W = 1.0 / spec.kW
        k_eps = 1.0 / spec.kEps

        N = int(ceil(T(k_eps) * K(k_W) / (K(k_eps) * T(k_W))))

        zeros, poles = EllipticFilter.get_normed_zeros_poles(N, spec.EpsP, spec.EpsS)

        z_poles = DigitalFilter.to_z_array(poles, spec.dt, spec.Wp)

        # z-нули из аналоговых нулей; нечётный порядок → добавляем z=-1
        z_zeros_list = list(DigitalFilter.to_z_array(zeros, spec.dt, spec.Wp))
        if N % 2 != 0:
            z_zeros_list = [-1.0 + 0j] + z_zeros_list
        z_zeros = np.array(z_zeros_list)

        # нормировка усиления на нулевой частоте (z=1)
        k_z = np.prod([1.0 - z for z in z_zeros])
        k_p = np.prod([1.0 - z for z in z_poles])
        k0 = 1.0 if N % 2 != 0 else (1.0 / sqrt(1.0 + spec.EpsP ** 2))
        g_norm = k0 * abs(k_p / k_z)

        B = np.poly(z_zeros).real * g_norm
        A = np.poly(z_poles).real
        return A, B

    def __init__(self, dt: float, fp: float, fs: float,
                 Gp: float = 0.891250938, Gs: float = 0.01):
        """
        Создаёт ФНЧ эллиптический.

        Параметры
        ---------
        dt : период дискретизации (с)
        fp : граничная частота полосы пропускания (Гц)
        fs : граничная частота полосы заграждения (Гц)
        Gp : коэффициент передачи на границе пропускания
        Gs : коэффициент передачи на границе заграждения
        """
        spec = Specification(dt, fp, fs, Gp, Gs)
        A, B = EllipticLowPass._initialize(spec)
        super().__init__(B, A, spec)
